pkg/v1/tkg/test/tkgctl/shared/e2e_tkrresolver.go

```diff
@@ -0,0 +1,358 @@
+		rand.Seed(time.Now().UnixNano())
+		mcClusterName := input.E2EConfig.ManagementClusterName
+		mcContextName = mcClusterName + "-admin@" + mcClusterName
+		mcProxy = framework.NewClusterProxy(mcClusterName, "", mcContextName)
+
+		logsDir = filepath.Join(input.ArtifactsFolder, "logs")
+
+		rand.Seed(time.Now().UnixNano())
+		clusterName = input.E2EConfig.ClusterPrefix + "wc-" + util.RandomString(4) // nolint:gomnd
+
+		tkgCtlClient, err = tkgctl.New(tkgctl.Options{
+			ConfigDir: input.E2EConfig.TkgConfigDir,
+			LogOptions: tkgctl.LoggingOptions{
+				File:      filepath.Join(logsDir, clusterName+".log"),
+				Verbosity: input.E2EConfig.TkgCliLogLevel,
+			},
+		})
+
+		Expect(err).To(BeNil())
+	})
+
+	It("Should create workload cluster with default kubernetes version and verify infra machine images are resolved correctly", func() {
+		By(fmt.Sprintf("Creating a workload cluster %q", clusterName))
+
+		options := framework.CreateClusterOptions{
+			ClusterName:  clusterName,
+			Namespace:    namespace,
+			Plan:         "dev",
+			CniType:      input.Cni,
+			OtherConfigs: input.OtherConfigs,
+		}
+		if input.Plan != "" {
+			options.Plan = input.Plan
+		}
+
+		if input.E2EConfig.InfrastructureName == "vsphere" {
+			if input.Cni == "antrea" {
+				if clusterIP, ok := os.LookupEnv("CLUSTER_ENDPOINT_ANTREA"); ok {
+					options.VsphereControlPlaneEndpoint = clusterIP
+				}
+			}
+			if input.Cni == "calico" {
+				if clusterIP, ok := os.LookupEnv("CLUSTER_ENDPOINT_CALICO"); ok {
+					options.VsphereControlPlaneEndpoint = clusterIP
+				}
+			}
+		}
+
+		clusterConfigFile, err := framework.GetTempClusterConfigFile(input.E2EConfig.TkgClusterConfigPath, &options)
+		Expect(err).To(BeNil())
+
+		defer os.Remove(clusterConfigFile)
+
+		err = tkgCtlClient.CreateCluster(tkgctl.CreateClusterOptions{
+			ClusterConfigFile: clusterConfigFile,
+			Edition:           "tkg",
+			Namespace:         namespace,
+		})
+		Expect(err).To(BeNil())
+
+		By(fmt.Sprintf("Validating the TKR data after cluster %q is created", clusterName))
+		verifyTKRData(context, mcProxy, options.ClusterName, options.Namespace)
+
+		By(fmt.Sprintf("Deleting workload cluster %q", clusterName))
+		err = tkgCtlClient.DeleteCluster(tkgctl.DeleteClustersOptions{
+			ClusterName: clusterName,
+			Namespace:   namespace,
+			SkipPrompt:  true,
+		})
+		Expect(err).To(BeNil())
+
+		By("Test successful !")
+	})
+
+	It("Should upgrade the cluster and verify infra machine images are resolved correctly", func() {
+		By(fmt.Sprintf("Creating a workload cluster %q", clusterName))
+
+		tkgBOMConfigClient := tkgconfigbom.New(input.E2EConfig.TkgConfigDir, nil)
+		defaultTKRVersion, err := tkgBOMConfigClient.GetDefaultTKRVersion()
+
+		Expect(err).ToNot(HaveOccurred(), "failed to get the default TKR version")
+
+		var defaultTKR, oldTKR *runv1.TanzuKubernetesRelease
+		Eventually(func() bool {
+			tkrs := mcProxy.GetTKRs(context)
+			defaultTKR, oldTKR = getTKRsForUpgrade(defaultTKRVersion, tkrs)
```

Comment:
Without the cluster, we have no idea if the TKRs will work for it: the cluster may not be able to use them (due to lack of OSImages satisfying the label selectors in CP and MDs).

Reply:
If we have an idea of what the cluster looks like, we can filter a list of TKRs by their applicability to the cluster (using a throwaway cluster.DeepCopy() in the filter func).